### PR TITLE
Add section starting Async guidelines

### PR DESCRIPTION
**Pull request by stevemchugh**
Initial commit for Async guidelines in csharp based on info from Alan. Needs a bit of technical writing adjustment...

### CODE REVIEW

File: coding/csharp.md [hunk 1]
```diff
@@ -289,6 +289,82 @@ public class MyObject
 }
 ```
 
+### Async
+
+If you ever need to write Task.Run or task.Wait or task.Result you are probably doing it wrong.
+
+Take this code:
```
Comment (tim-crowther) [inline coding/csharp.md]: "For example;"

File: coding/csharp.md [hunk 2]
```diff
@@ -289,6 +289,82 @@ public class MyObject
...
+
+            return CreateStreamDetails(blob, filename);
+        });
+
+    return await task;
+}
+```
+What it is doing is starting a task on a separate thread (Task.Run) from the thread pool.
+
+Then that thread gets to the time expensive operations and is blocked:
```
Comment (tim-crowther) [inline coding/csharp.md]: "This leads to the task startong on a separate thread (Task.Run) from the thread pool. Then that thread reaches the time expensive operations and is blocked:
"

File: coding/csharp.md [hunk 3]
```diff
@@ -289,6 +289,82 @@ public class MyObject
...
+  IBlobWrapper blob = _azureWrapper.GetFileBlob(filename, PublishAssetContainer, path);
+  return CreateStreamDetails(blob, filename);
+
+While it is blocked, that thread will not be returned to the pool to do other work.
+
```
Comment (tim-crowther) [inline coding/csharp.md]: Delete whitespace here.

File: coding/csharp.md [hunk 4]
```diff
@@ -289,6 +289,82 @@ public class MyObject
...
+The number of threads in the threadpool is limited, increasing the limit is possible, but it is impossible to determine a good number to pick.
+
```
Comment (tim-crowther) [inline coding/csharp.md]: May be worth referencing this https://msdn.microsoft.com/en-us/library/0ka9477y(v=vs.110).aspx

File: coding/csharp.md (hunk 3 above)
Comment (stevemchugh) [inline coding/csharp.md]: Done

File: coding/csharp.md (hunk 1 above)
Comment (stevemchugh) [inline coding/csharp.md]: Done

File: coding/csharp.md [hunk 5]
```diff
@@ -289,6 +289,82 @@ public class MyObject
 }
 ```
 
+### Async
+
+If you ever need to write Task.Run or task.Wait or task.Result you are probably doing it wrong.
```
Comment (stevemchugh) [inline coding/csharp.md]: Done

File: coding/csharp.md [hunk 6]
```diff
@@ -289,6 +289,82 @@ public class MyObject
...
+    };
+
+    streamDetails.Stream.Position = 0;
+
+    return streamDetails;
+}
+```
+
+Now the thread will be returned to the pool while it awaits the non-CPU bound operations, e.g. in this case connecting over the network to azure. This maximizes the work the CPU cores can do without increasing thread count.
+
```
Comment (stevemchugh) [inline coding/csharp.md]: This is specifically related to Async scenarios and examples, the link is to a more general discussion on threading and my confuse in this instance. Maybe reserve for a more general discussion on threading...

**Pull request merged**